src/appleseed-max-impl/appleseedoslplugin/oslparamdlg.cpp

```diff
@@ -0,0 +1,260 @@
+        dialog_template.AddComponent((LPCSTR)"SpinnerControl", (LPCSTR)"Spinner", WS_VISIBLE | WS_TABSTOP, NULL, m_col_2 + m_edit_width + 2, y_pos, 7, m_edit_height, ctrl_id++);
+        dialog_template.AddComponent((LPCSTR)"CustEdit", (LPCSTR)"Edit", WS_VISIBLE | WS_TABSTOP, NULL, m_col_3, y_pos, m_edit_width, m_edit_height, ctrl_id++);
+        dialog_template.AddComponent((LPCSTR)"SpinnerControl", (LPCSTR)"Spinner", WS_VISIBLE | WS_TABSTOP, NULL, m_col_3 + m_edit_width + 2, y_pos, 7, m_edit_height, ctrl_id++);
+        dialog_template.AddComponent((LPCSTR)"CustEdit", (LPCSTR)"Edit", WS_VISIBLE | WS_TABSTOP, NULL, m_col_4, y_pos, m_edit_width, m_edit_height, ctrl_id++);
+        dialog_template.AddComponent((LPCSTR)"SpinnerControl", (LPCSTR)"Spinner", WS_VISIBLE | WS_TABSTOP, NULL, m_col_4 + m_edit_width + 2, y_pos, 7, m_edit_height, ctrl_id++);
+        break;
+      case MaxParam::IntCheckbox:
+        dialog_template.AddButton((LPCSTR)"", WS_VISIBLE | BS_AUTOCHECKBOX | WS_TABSTOP, NULL, m_col_2, y_pos, 10, 10, ctrl_id++);
+        break;
+      case MaxParam::IntMapper:
+      case MaxParam::StringPopup:
+        dialog_template.AddComboBox((LPCSTR)"", CBS_DROPDOWNLIST | WS_VSCROLL | WS_TABSTOP | WS_VISIBLE, NULL, m_col_2, y_pos, m_tex_width, m_edit_height, ctrl_id++);
+        y_pos += 2;
+        break;
+      case MaxParam::Color:
+        dialog_template.AddComponent((LPCSTR)"ColorSwatch", (LPCSTR)"Color", WS_VISIBLE | WS_TABSTOP, NULL, m_col_2, y_pos, m_edit_width, m_edit_height, ctrl_id++);
+        break;
+      case MaxParam::String:
+        dialog_template.AddComponent((LPCSTR)"CustEdit", (LPCSTR)"Edit", WS_VISIBLE | WS_TABSTOP, NULL, m_col_2, y_pos, m_tex_width, m_edit_height, ctrl_id++);
+        break;
+      case MaxParam::Closure:
+        dialog_template.AddComponent((LPCSTR)"CustButton", (LPCSTR)"", WS_VISIBLE, NULL, m_col_3, y_pos, m_tex_width, m_edit_height, ctrl_id++);
+    }
+
+    if (max_param.connectable && max_param.param_type != MaxParam::Closure)
+    {
+        Texmap* tex_map = nullptr;
+        Interval iv;
+        m_osl_plugin->GetParamBlock(0)->GetValue(max_param.max_param_id + 1, 0, tex_map, iv);
+        
+        const char* label = tex_map != nullptr ? "M" : "";
+
+        if (max_param.is_vector())
+            dialog_template.AddComponent((LPCSTR)"CustButton", (LPCSTR)label, WS_VISIBLE, NULL, m_col_4 + m_edit_width + 11, y_pos, m_edit_height, m_edit_height, ctrl_id);
+        else
+            dialog_template.AddComponent((LPCSTR)"CustButton", (LPCSTR)"", WS_VISIBLE, NULL, m_col_3, y_pos, m_tex_width, m_edit_height, ctrl_id);
+    }
+
+    y_pos += 12;
+    if (param_page != pages.end())
+        param_page->second.y_pos = y_pos;
+}
+
+void OSLParamDlg::create_dialog()
+{
+    PageGroupMap pages;
+
+    for (auto& osl_param : m_shader_info->m_params)
+    {
+        if (osl_param.page != "" &&
+            osl_param.page != "UV Coordinates" &&
```

Comment:
Isn't there some sort of stable id that we could use instead of relying on a label (which could change, be translated...)?

Reply:
UV coordinates parameters are exposed in Maya ui, but they are useless in Max, since UV parameters are connected inside the plugin. That's why I decided to hide them this way.